### PR TITLE
Add restoreBackup support for sql db instance

### PR DESCRIPTION
**Pull request by c2thorn**
Fixes https://github.com/hashicorp/terraform-provider-google/issues/2446

Add support for the very imperative [restoreBackup](https://cloud.google.com/sql/docs/mysql/admin-api/rest/v1beta4/instances/restoreBackup) action. This change essentially triggers the restoreBackup action:

1. After the resource is created and `restore_backup_context` is set
2. At the end of update if `restore_backup_context` has a change

If this PR is for Terraform, I acknowledge that I have:

- [X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ ] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [ ] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [X] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
sql: added restore from backup support to `google_sql_database_instance`
```

### CODE REVIEW

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 1]
```diff
@@ -1025,6 +1056,8 @@ func resourceSqlDatabaseInstanceUpdate(d *schema.ResourceData, meta interface{})
 		return err
 	}
 
+	// TODO if d.HasChange("settings") doesn't work, so we just send an empty request if only restoring from backup
```
Comment (c2thorn) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb:1059]: Thanks to https://github.com/hashicorp/terraform-plugin-sdk/issues/98 and `settings.authorized_networks`, d.HasChange("settings") always returns true.

As it stands the resource will send empty updates which causes an extra harmless request.

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 2]
```diff
@@ -1051,6 +1084,15 @@ func resourceSqlDatabaseInstanceUpdate(d *schema.ResourceData, meta interface{})
 		return err
 	}
 
+	if r, ok := d.GetOk("restore_backup_context"); ok {
```
Comment (c2thorn) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb:1087]: To support restoring from a backup during Update, we'll have to trigger when changes are made to `restore_backup_context` other than the field just being removed

**Comment by c2thorn** [comment]
The draft is still pending:

Tests showcasing create + update restoreBackup triggers
Documentation

**Review by rileykarson — COMMENTED** [review]
This seems like a reasonable take. We'll want to be explicit in the documentation that this is a pretty dangerous field to set & document how it may behave differently than other fields.

**Comment by c2thorn** [comment]
Test capability is pending https://github.com/GoogleCloudPlatform/magic-modules/pull/4352 for datasource use and backup bootstrapping

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 2 above)
Comment (rileykarson) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb:1085]: Can you comment that inline?

File: third_party/terraform/resources/resource_sql_database_instance.go.erb [hunk 3]
```diff
@@ -583,6 +583,30 @@ settings.backup_configuration.binary_log_enabled are both set to true.`,
...
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"backup_run_id": {
+							Type:        schema.TypeInt,
+							Required:    true,
+							Description: `The ID of the backup run to restore from.`,
+						},
+						"instance_id": {
```
Comment (rileykarson) [inline third_party/terraform/resources/resource_sql_database_instance.go.erb:597]: Just curious: what does specifying this field (or not specifying this field) change?

**Review by rileykarson — APPROVED** [review]

File: third_party/terraform/resources/resource_sql_database_instance.go.erb (hunk 3 above)
Comment (c2thorn) [reply third_party/terraform/resources/resource_sql_database_instance.go.erb:597]: You can optionally restore to the backup of another instance (backup ID's are unique per instance). If left blank, the API just assumes the same instance. I can add a line in the documentation to make this more clear.

**Review by rileykarson — APPROVED** [review]

**Comment by achew22** [comment]
@rileykarson and @c2thorn, I believe this has caused a regression. I was able to trigger it by `terraform import`ing a different resource (in my case a GCS bucket) using the follwoing:

```sh
$ terraform import module.modulename.google_storage_bucket.bucketname bucketname
module.modulename.google_storage_bucket.bucketname: Importing from ID "bucketname"...
module.modulename.google_storage_bucket.bucketname: Import prepared!
  Prepared google_storage_bucket for import
module.modulename.google_storage_bucket.bucketname: Refreshing state... [id=bucketname]

Error: Invalid resource instance data in state

  on ../modules/sql/main.tf line 19:
  19: resource "google_sql_database_instance" "postgres" {

Instance module.sql.google_sql_database_instance.postgres
data could not be decoded from the state: unsupported attribute
"restore_backup_context".
```

This does not happen when I run `terraform apply`, only when I try to import another resource.

```sh
$ terraform version
Terraform v0.14.5
+ provider registry.terraform.io/hashicorp/aws v3.24.1
+ provider registry.terraform.io/hashicorp/google v3.52.0
+ provider registry.terraform.io/hashicorp/google-beta v3.52.0
+ provider registry.terraform.io/hashicorp/kubernetes v1.13.3
+ provider registry.terraform.io/hashicorp/kubernetes-alpha v0.2.1
+ provider registry.terraform.io/hashicorp/random v3.0.1
+ provider registry.terraform.io/hashicorp/tls v3.0.0
```

**Comment by c2thorn** [comment]
Hey @achew22, sorry for any troubles that this may have caused you. I tried reproducing this error myself by importing a storage bucket while already having a database instance in state, but was not able to get the error.

Do you mind filing a new issue with the steps you have taken and potentially your terraform configuration that produced this error?

**Comment by achew22** [comment]
@c2thorn I would be very happy to do this. Do you have any hints on things that you think might be related to this. This is deep in the bowels of our terraform config and it even took me a while to get the stack trace (not sure what else to call it) cleaned up for publication. I can poke at it, but I'm, unfortunately timeboxed on this one.

**Comment by c2thorn** [comment]
@achew22 I just realized that this feature was released in [v3.53](https://github.com/hashicorp/terraform-provider-google/releases/tag/v3.53.0) of the providers and your log shows v3.52. This seems like a versioning issue where maybe some part of your config is using a different version (`google_storage_bucket` maybe?)

Maybe you can try constraining 3.53 or newer via https://www.terraform.io/docs/configuration/provider-requirements.html#version-constraints ?

**Comment by achew22** [comment]
@c2thorn Sorry to raise the false alarm. I think you're right. I was able to force my local terraform to v3.53 and it seems to be working now. Thanks so much!

**Comment by c2thorn** [comment]
@achew22 Great! Glad you were able to get unstuck!

**Pull request merged**